### PR TITLE
fix(compiler): throw for hierarchical component selectors

### LINKED ISSUE #2589: Non-simple selectors should throw error
```
@Directive({
  selector: 'parent child'
})
```

Should throw an error, since we don't support selectors which cross elements.

### LINKED ISSUE #6600: Invalid component selectors should throw an error
If you use an unsupported component selector, such as `parent-thing > child-thing`, Angular will silently accept this and just capture `child-thing`.

### PR DESCRIPTION
**Pull request by pkozlowski-opensource**
Fixes #6600
Fixes #2589

It is potentially a breaking change since previously hierarchical component selectors (ex. `taga tagb`) would match `tagb` (regardless of a parent) and some people might, accidentally have their app working. Let me know if this one should be marked as a BC.

### CODE REVIEW

**Comment by pkozlowski-opensource** [comment]
HMM, it looks like we are using `CssSelector` in parse5 adapter (huh? I thought it was only for component selectors!) hence the failing tests. Will look into it.

File: modules/@angular/compiler/src/selector.ts [hunk 1]
```diff
@@ -7,12 +7,13 @@ const _EMPTY_ATTR_VALUE = /*@ts2dart_const*/ '';
...
-    '([-\\w]+)|' +                            // "tag"
-    '(?:\\.([-\\w]+))|' +                     // ".class"
-    '(?:\\[([-\\w*]+)(?:=([^\\]]*))?\\])|' +  // "[name]", "[name=value]" or "[name*=value]"
-    '(\\))|' +                                // ")"
-    '(\\s*,\\s*)');                           // ","
+    '(\\:not\\()|' +                           //":not("
+    '([-\\w]+)|' +                             // "tag"
+    '(?:\\.([-\\w]+))|' +                      // ".class"
+    '(?:\\[([-\\w*^]+)(?:=([^\\]]*))?\\])|' +  // "[name]", "[name=value]" or "[name*=value]" or
+                                               // "[name^=value]"
```
Comment (pkozlowski-opensource) [inline modules/@angular/compiler/src/selector.ts]: `"[name^=value]"` "support" here is a very, very lame hack to not break existing tests with parse5 adapter. The issue is that parse5 adapter [is using](https://github.com/angular/angular/blob/2d60ff14aefdf18a939e4d65b0856b90c376e054/modules/%40angular/platform-server/src/parse5_adapter.ts#L108) this class as a general-purpose selection handling. We can't have it both ways - either we use this class for component selectors only (and then we can have good error messages) or as a general purpose selection utility (and then we can't restrict what is available for component selector).

**Comment by pkozlowski-opensource** [comment]
tl;dr; This is a hacky "fix". We need to decide what is the role / purpose of the `CssSelector` class is before we can have a proper fix.

The issue is that parse5 adapter [is using](https://github.com/angular/angular/blob/2d60ff14aefdf18a939e4d65b0856b90c376e054/modules/%40angular/platform-server/src/parse5_adapter.ts#L108) this class as a general-purpose selection handling. We can't have it both ways - either we use this class for component selectors only (and then we can have good error messages) or as a general purpose selection utility (and then we can't restrict what is available for component selector).

**Comment by pkozlowski-opensource** [comment]
My vote goes to removing usage of `CssSelector` from `parse5_adapter` as the `CssSelector` is rather limited and can't be used a general-purpose selection utility. The other thing to explore is selector-related code sharing with the existing CSS parser.

**Comment by vicb** [comment]
What is the status here ?

**Comment by pkozlowski-opensource** [comment]
> What is the status here ?

@vicb to progress on this one I would need help / input on https://github.com/angular/angular/pull/9347#issuecomment-226988823

**Comment by tbosch** [comment]
@pkozlowski-opensource I agree, we should not use the `CssSelector` in the parse5 adapter. The `CssSelector` should be optimized for error reporting in templates.

Could you change parse5 adapter to use its own matching logic?

**Comment by pkozlowski-opensource** [comment]
> Could you change parse5 adapter to use its own matching logic?

I will. Then it will be easier to report errors, optimize etc. 

Going to close this PR for now, will open separate ones for parse5 adapter and CSS selector for directives matching. Thnx for the feedback @tbosch !
